Refactor Spree to make MOSS taxation possible

mamhoff opened the pull request:
1.  Add current_tax_zone helper
   
   The `current_tax_zone` is needed for displaying the correct price
   including VAT.
   
   In order to cache product prices that depend on the current tax zone, 
   we need to add the current tax zone to the cache key for a single or
   multiple products.
2.  Display correct VAT prices
   1. Add Spree::Price#price_including_tax_for(zone)
      
      This methods returns the correctly calculated gross price
      for any zone that might have VAT. This way we can make sure to
      always include the correct VAT in prices (even if they differ).
      
      The method is delegated to Spree::Product and Spree::Variant.
   2. Amend display_price helper to use current tax zone
3. Copy price including VAT from Variant
   
   This enables us to always have the same price as displayed
   in the order. It also frees us from having to load default VAT
   rates in `tax_rate.rb`.
   
   The line item price will only ever be updated after the address
   step in the state machine.
   
   One expectation is changed in the back-end, because here the
   price of the line items is updated from the variant, and those are
   different.
4. Refactor Spree::Zone and Spree::TaxRate
   
   When adjusting line items, do not include the default rates
   in the Array of applicable rates. This was only necessary because
   we could not be sure about the price being correct in the first place,
   which is now fixed through ´Spree::Price#price_inlcuding_vat_for(zone)´.
   
   As a consequence, I could refactor the stock estimator and fix
   https://github.com/spree/spree/issues/6234.
5. Add rounding spec to tax_rate_spec.rb
   
   This is mainly for posterity: I got stuck in rounding hell once, and
   this test should ensure no-one else will have to.
6. Fix reimbursement expectations
   
   The reimbursement spec expected negative adjustments for included tax,
   which aren't necessary anymore - if your local tax zone does not include
   the default VAT zone's VAT, the price is adjusted accordingly
   beforehand.
7. Add test for MOSS taxes
   
   This is an elaborate almost-integration-test for a particular setup
   proving that it actually _is_ possible now to configure Spree for MOSS.

mamhoff committed:
Refactor Spree to make MOSS taxation possible

1. Add current_tax_zone helper

The `current_tax_zone` is needed for displaying the correct price
including VAT.

In order to cache product prices that depend on the current tax zone,
we need to add the current tax zone to the cache key for a single or
multiple products.

2. Display correct VAT prices

2. 1. Add Spree::Price#price_including_tax_for(zone)

This methods returns the correctly calculated gross price
for any zone that might have VAT. This way we can make sure to
always include the correct VAT in prices (even if they differ).

The method is delegated to Spree::Product and Spree::Variant.

2. 2. Amend display_price helper to use current tax zone

3. Copy price including VAT from Variant

This enables us to always have the same price as displayed
in the order. It also frees us from having to load default VAT
rates in `tax_rate.rb`.

The line item price will only ever be updated after the address
step in the state machine.

One expectation is changed in the back-end, because here the
price of the line items is updated from the variant, and those are
different.

4. Refactor Spree::Zone and Spree::TaxRate

When adjusting line items, do not include the default rates
in the Array of applicable rates. This was only necessary because
we could not be sure about the price being correct in the first place,
which is now fixed through ´Spree::Price#price_inlcuding_vat_for(zone)´.

As a consequence, I could refactor the stock estimator and fix
https://github.com/spree/spree/issues/6234.

5. Add rounding spec to tax_rate_spec.rb

This is mainly for posterity: I got stuck in rounding hell once, and
this test should ensure no-one else will have to.

6. Fix reimbursement expectations

The reimbursement spec expected negative adjustments for included tax,
which aren't necessary anymore - if your local tax zone does not include
the default VAT zone's VAT, the price is adjusted accordingly
beforehand.

7. Add test for MOSS taxes

This is an elaborate almost-integration-test for a particular setup
proving that it actually *is* possible now to configure Spree for MOSS.

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://magiclabs-refactor-tax-on-master.spree-spree-9ed724b77adec04c49c2.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://magiclabs-refactor-tax-on-master.spree-spree-9ed724b77adec04c49c2.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

mamhoff committed:
Add documentation for the new VAT handling

* Release notes
* Adapt taxation guide

It also includes some of the corrections mentioned by members of the
community.

mamhoff commented:
Hm, this PR used to be green until I added documentation, at which point random feature specs turn red. Should I keep pushing until it's green?

JDutil commented:
Thanks for the continued effort on this.  Browsing through it things look pretty good, but I'd really like some other VAT users to review as well like they did on the previous PR. 

/cc @alepore @kennyadsl @simmsy @forvalho @paulo-surfdome you guys would also be interested in this.

kennyadsl commented:
@mamhoff it seems perfect to me. Thanks for this!

JDutil commented on backend/spec/features/admin/orders/line_items_spec.rb:
```diff
@@ -27,7 +27,7 @@
         end
         within '.line-item-total' do
-          expect(page).to have_content("$100.00")
+          expect(page).to have_content("$199.90")
```
why does this price change I was thinking non-vat would be handled the same?

mamhoff replied:
It is. It's just that in this test, the test setup sets up an order with line item prices of 10, and the line item prices are overwritten by the variant that is also created. The default variant price is 19.99. Since the spec is about multiplying, I think that's actually okay :).

JDutil replied:
:+1: makes sense I ran into a similar spec issue recently.  Not sure why we don't default line item factory to be same price as variant.

JDutil commented on core/spec/models/spree/shipping_rate_spec.rb:
```diff
@@ -38,20 +38,21 @@
 
-        it "shows correct tax amount" do
-          expect(shipping_rate.display_price.to_s).to eq("$10.00 (excl. $0.91 #{tax_rate.name})")
+        pending "I would expect this to be analogue to line items: First calculate the net price, then add the foreign VAT.\n"+
```
we should resolve this before merging

BenMorganIO replied:
agreed

mamhoff replied:
Bear in mind that this is the edge case for charging shipping costs for downloads. Anyone ever paid for a being able to download something, rather than have it shipped?

BenMorganIO commented on core/app/models/spree/tax_rate.rb:
```diff
@@ -16,51 +6,31 @@ class TaxRate < Spree::Base
+    scope :by_zone, -> (zone) { where(zone_id: zone.id) }
+    scope :potential_rates_for_zone,
+      -> (zone) do
+        where(zone_id: Spree::Zone.potential_matching_zones(zone).pluck(:id))
```
I can see this possibly cause an n+1. How is this scope used?

mamhoff replied:
It replaces the method deleted above (which I think would cause the same n+1 if so). It is only called from within tax_rate.rb (is there such a thing as private scopes?), in order to match the zone of the tax rate with the zone of the order. This happens often: 
- Every time the order is updated
- Every time a price has to be displayed for an order with a tax zone outside the default zone.

BenMorganIO commented on core/app/models/spree/tax_rate.rb:
```diff
@@ -16,51 +6,31 @@ class TaxRate < Spree::Base
+        where(zone_id: Spree::Zone.potential_matching_zones(zone).pluck(:id))
+      end
+    scope :for_default_zone,
+      -> { potential_rates_for_zone(Spree::Zone.default_tax) }
```
This one as well. Is there a way in rails to associate over to the zone model and call `zone.class.default_tax`?

BenMorganIO commented on core/lib/spree/core/controller_helpers/store.rb:
```diff
@@ -16,6 +17,14 @@ def current_currency
+          else
+            Spree::Zone.default_tax
+          end
+        end
```
```ruby
current_order.try(:tax_zone) || Spree::Zone.default_tax
```

mamhoff replied:
Yep

BenMorganIO commented on core/spec/helpers/base_helper_spec.rb:
```diff
@@ -134,4 +134,71 @@ def link_to_tracking_html(options = {})
+
+  describe '#display_price' do
+    let!(:product) { create(:product) }
+    let(:current_currency) { "USD" }
```
Isn't the default currency already USD?

mamhoff replied:
Yes, but in the test context the helper method is not defined.

BenMorganIO commented on core/spec/models/spree/reimbursement_spec.rb:
```diff
@@ -118,7 +118,7 @@
         }.to change { Spree::Refund.count }.by(1)
         return_item.reload
-        expect(return_item.included_tax_total).to be < 0
+        expect(return_item.included_tax_total).to be > 0
```
Can you explain this one? I think a reimbursement should be negative.

mamhoff replied:
Reimbursements are positive; the negative included tax adjustment would be (I think) for a tax refund reimbursement. Since now there is no tax refunds anymore as the price is right from the start, the refund doesn't need to be reimbursed.

BenMorganIO commented on guides/content/developer/core/taxation.md:
```diff
@@ -105,16 +105,12 @@ Finally, if the taxable address (either the shipping or billing, depending on th
 Many jurisdictions have what is commonly referred to as a Value Added Tax (VAT.) In these cases the tax is typically applied to the price. This means that prices for items are "inclusive of tax" and no additional tax needs to be applied during checkout.
 
-In the case of tax inclusive pricing the store owner should enter all prices inclusive of tax. This makes it easy for Spree to display the tax inclusive price since it won't have to be constantly calculated on the fly.
+In the case of tax inclusive pricing the store owner can enter all prices inclusive of tax if she has set her home country as the default zone. If there is no default zone set, any taxes that are included in the price will be added to the net prices on the fly depending on the current order's tax zone.
```
"if their home country is the default zone"
What is "on the fly"? When does this happen in the process of checkout?

mamhoff replied:
Okay for 'their'. 

When the current order already has a zone that differs from default zone (empty cart, address entered): 
There is a `Spree::Price#price_including_vat_for(zone)` method, which does that calculation - both on [display](https://github.com/magiclabs/spree/blob/refactor-tax-on-master/core/app/helpers/spree/base_helper.rb#L19-L21) and on [creating](https://github.com/magiclabs/spree/blob/refactor-tax-on-master/core/app/models/spree/line_item.rb#L13) [line items](https://github.com/magiclabs/spree/blob/refactor-tax-on-master/core/app/models/spree/line_item.rb#L37-L49). 

The prices are also updated [after the address step](https://github.com/magiclabs/spree/blob/refactor-tax-on-master/core/app/models/spree/order/checkout.rb#L90).

BenMorganIO commented on guides/content/developer/core/taxation.md:
```diff
@@ -133,6 +129,25 @@ Finally, if the order's address is changed to being outside this tax zone, then
+
+#### Differing VATs for different product categories depending on the customer's zone
+
+As of January 1st, 2015, digital products sold within the EU have to have the VAT of the receiving country applied. Physical products have to have the seller's VAT applied. In order to set this up, please proceed as follows:
```
"have to have" should be "Physical products must have the seller's VAT applied."

http://english.stackexchange.com/questions/35545/is-this-correct-id-have-to-have-had

mamhoff replied:
Haha :) thanks.

BenMorganIO commented on guides/content/developer/core/taxation.md:
```diff
@@ -133,6 +129,25 @@ Finally, if the order's address is changed to being outside this tax zone, then
+
+2. Mark your home zone as the default zone so you can conveniently enter gross prices.
+
+3. Create a tax category "Digital products", and a tax category "Physical products"
```
Missing period.

mamhoff replied:
Fixed.

BenMorganIO commented:
For the record I actually like your grammar in the documentation copy. :+1:

mamhoff committed:
Refactor calculating gross prices and shipping costs

Use the same logic for shipping rates as for line items: Upon creation
of the cost, apply the relevant included tax rates.

This resulted in some duplicate code, which is now extracted into a
concern and tested for not calling the database when it does not have
to.

Move calculating of gross amounts into a concern and test the concern
for calling `TaxRate.included_tax_amount_for` as little as possible.

mamhoff commented:
I talked to our tax advisor, and things are as I supposed: Shipment costs should be taxes like their respective line item costs. In order to have this work analogous to line items, I adjust the price in `Spree::Stock::Estimator`.

alepore commented:
:+1: thanks again for the this work

alepore commented on core/app/helpers/spree/products_helper.rb:
```diff
@@ -54,7 +54,11 @@ def line_item_description_text description_text
+    end
+
+    def cache_key_for_product(product = @product)
+      "#{I18n.locale}/#{current_currency}/#{current_tax_zone.try(:cache_key)}/#{product.cache_key}"
```
(minor stuff) maybe using an array here is more clean?
```ruby
[I18n.locale, current_currency, current_tax_zone, product]
```

mamhoff replied:
Yes, nice one!

mamhoff commented:
Hound is happy. Finally. Thank you @alepore, @kennyadsl  and @BenMorganIO for your helpful comments. 

I've also moved the MOSS taxes spec to `tax_rate_spec.rb`, as it really just tests one method in there. 

We've also refactored some of the labeling code for adjustments so they can be translated more easily. There's a couple of additional I18n keys to sync now - once this PR is through, we'll add those to the spree_i18n gem for the languages we speak. 

@JDutil anything else I can do?

JDutil commented on core/app/models/spree/order.rb:
```diff
@@ -277,6 +277,13 @@ def create_tax_charge!
+    def update_line_item_prices!
+      line_items.each do |line_item|
+        line_item.update_price
+        line_item.save
```
I wonder if we could improve this to save all line items at once to try squeezing performance gains here.  Quick google search isn't turning up much in the way of batching these records, but what if we do a single database transaction:
http://stackoverflow.com/a/2510258

Not sure how much better it would be, but I imagine it could add up for orders with many line items.

JDutil commented:
Great work @mamhoff everything looks good to me now except for the last comment I made about possibly using transaction block when updating the line item prices.  My only concern here is that adding yet more callbacks could degrade performance, but I don't see a better alternative.  I'd be curious to see any benchmark or query differences on the transition from address before/after a transaction block there.

mamhoff committed:
Wrap updating line items in a transaction

As @JDutil noted in
https://github.com/spree/spree/pull/6295#discussion_r28899769

Thanks for the suggestion! :)

mamhoff commented:
Oh Hound, I love you. 

I can't provide meaningful benchmarks, but I can show some SQL queries. It's definitely faster with the transaction: 
``` ruby 
o = Spree::Order.first
o.transaction { o.line_items.map(&:update_price); o.save }
```
``` sql
   (0.3ms)  BEGIN
  Spree::LineItem Load (1.0ms)  SELECT "spree_line_items".* FROM "spree_line_items" WHERE "spree_line_items"."order_id" = $1  ORDER BY spree_line_items.created_at ASC  [["order_id", 1]]
  Spree::Variant Load (1.2ms)  SELECT  "spree_variants".* FROM "spree_variants" WHERE "spree_variants"."id" = $1 LIMIT 1  [["id", 18]]
  Spree::Zone Load (0.6ms)  SELECT  "spree_zones".* FROM "spree_zones" WHERE "spree_zones"."default_tax" = 't'  ORDER BY "spree_zones"."id" ASC LIMIT 1
  Spree::Price Load (0.7ms)  SELECT  "spree_prices".* FROM "spree_prices" WHERE "spree_prices"."variant_id" = $1 AND "spree_prices"."currency" = $2 LIMIT 1  [["variant_id", 18], ["currency", "EUR"]]
  Spree::Variant Load (0.6ms)  SELECT  "spree_variants".* FROM "spree_variants" WHERE "spree_variants"."id" = $1 LIMIT 1  [["id", 18]]
  Spree::Product Load (1.0ms)  SELECT  "spree_products".* FROM "spree_products" WHERE "spree_products"."id" = $1 LIMIT 1  [["id", 18]]
  Spree::TaxCategory Load (0.9ms)  SELECT  "spree_tax_categories".* FROM "spree_tax_categories" WHERE "spree_tax_categories"."deleted_at" IS NULL AND "spree_tax_categories"."id" = $1 LIMIT 1  [["id", 1]]
  Spree::Zone Load (0.8ms)  SELECT  "spree_zones".* FROM "spree_zones" WHERE "spree_zones"."default_tax" = 't'  ORDER BY "spree_zones"."id" ASC LIMIT 1
  Spree::Variant Load (0.6ms)  SELECT  "spree_variants".* FROM "spree_variants" WHERE "spree_variants"."id" = $1 LIMIT 1  [["id", 19]]
  Spree::Price Load (0.6ms)  SELECT  "spree_prices".* FROM "spree_prices" WHERE "spree_prices"."variant_id" = $1 AND "spree_prices"."currency" = $2 LIMIT 1  [["variant_id", 19], ["currency", "EUR"]]
  Spree::Variant Load (0.6ms)  SELECT  "spree_variants".* FROM "spree_variants" WHERE "spree_variants"."id" = $1 LIMIT 1  [["id", 19]]
  Spree::Product Load (0.5ms)  SELECT  "spree_products".* FROM "spree_products" WHERE "spree_products"."id" = $1 LIMIT 1  [["id", 19]]
  Spree::TaxCategory Load (0.5ms)  SELECT  "spree_tax_categories".* FROM "spree_tax_categories" WHERE "spree_tax_categories"."deleted_at" IS NULL AND "spree_tax_categories"."id" = $1 LIMIT 1  [["id", 1]]
  Spree::Zone Load (0.6ms)  SELECT  "spree_zones".* FROM "spree_zones" WHERE "spree_zones"."default_tax" = 't'  ORDER BY "spree_zones"."id" ASC LIMIT 1
  Spree::Order Exists (0.9ms)  SELECT  1 AS one FROM "spree_orders" WHERE ("spree_orders"."number" = 'R051917678' AND "spree_orders"."id" != 1) LIMIT 1
   (0.4ms)  COMMIT
```
Without the transaction block: 

``` ruby
o = Spree::Order.first
o.update_line_item_prices! # old implementation
```
``` sql
Spree::LineItem Load (0.9ms)  SELECT "spree_line_items".* FROM "spree_line_items" WHERE "spree_line_items"."order_id" = $1  ORDER BY spree_line_items.created_at ASC  [["order_id", 1]]
  Spree::Variant Load (1.1ms)  SELECT  "spree_variants".* FROM "spree_variants" WHERE "spree_variants"."id" = $1 LIMIT 1  [["id", 18]]
  Spree::Zone Load (1.5ms)  SELECT  "spree_zones".* FROM "spree_zones" WHERE "spree_zones"."default_tax" = 't'  ORDER BY "spree_zones"."id" ASC LIMIT 1
  Spree::Price Load (0.8ms)  SELECT  "spree_prices".* FROM "spree_prices" WHERE "spree_prices"."variant_id" = $1 AND "spree_prices"."currency" = $2 LIMIT 1  [["variant_id", 18], ["currency", "EUR"]]
  Spree::Variant Load (0.7ms)  SELECT  "spree_variants".* FROM "spree_variants" WHERE "spree_variants"."id" = $1 LIMIT 1  [["id", 18]]
  Spree::Product Load (1.5ms)  SELECT  "spree_products".* FROM "spree_products" WHERE "spree_products"."id" = $1 LIMIT 1  [["id", 18]]
  Spree::TaxCategory Load (0.8ms)  SELECT  "spree_tax_categories".* FROM "spree_tax_categories" WHERE "spree_tax_categories"."deleted_at" IS NULL AND "spree_tax_categories"."id" = $1 LIMIT 1  [["id", 1]]
  Spree::Zone Load (0.9ms)  SELECT  "spree_zones".* FROM "spree_zones" WHERE "spree_zones"."default_tax" = 't'  ORDER BY "spree_zones"."id" ASC LIMIT 1
   (0.2ms)  BEGIN
  Spree::Product Load (0.6ms)  SELECT  "spree_products".* FROM "spree_products" WHERE "spree_products"."id" = $1 LIMIT 1  [["id", 18]]
  Spree::TaxCategory Load (0.5ms)  SELECT  "spree_tax_categories".* FROM "spree_tax_categories" WHERE "spree_tax_categories"."deleted_at" IS NULL AND "spree_tax_categories"."id" = $1 LIMIT 1  [["id", 1]]
   (0.9ms)  SELECT COUNT(*) FROM "spree_inventory_units" WHERE "spree_inventory_units"."line_item_id" = $1  [["line_item_id", 1]]
   (0.3ms)  COMMIT
  Spree::Variant Load (0.6ms)  SELECT  "spree_variants".* FROM "spree_variants" WHERE "spree_variants"."id" = $1 LIMIT 1  [["id", 19]]
  Spree::Price Load (0.5ms)  SELECT  "spree_prices".* FROM "spree_prices" WHERE "spree_prices"."variant_id" = $1 AND "spree_prices"."currency" = $2 LIMIT 1  [["variant_id", 19], ["currency", "EUR"]]
  Spree::Variant Load (0.3ms)  SELECT  "spree_variants".* FROM "spree_variants" WHERE "spree_variants"."id" = $1 LIMIT 1  [["id", 19]]
  Spree::Product Load (0.5ms)  SELECT  "spree_products".* FROM "spree_products" WHERE "spree_products"."id" = $1 LIMIT 1  [["id", 19]]
  Spree::TaxCategory Load (0.5ms)  SELECT  "spree_tax_categories".* FROM "spree_tax_categories" WHERE "spree_tax_categories"."deleted_at" IS NULL AND "spree_tax_categories"."id" = $1 LIMIT 1  [["id", 1]]
  Spree::Zone Load (0.7ms)  SELECT  "spree_zones".* FROM "spree_zones" WHERE "spree_zones"."default_tax" = 't'  ORDER BY "spree_zones"."id" ASC LIMIT 1
   (0.3ms)  BEGIN
  Spree::Product Load (0.5ms)  SELECT  "spree_products".* FROM "spree_products" WHERE "spree_products"."id" = $1 LIMIT 1  [["id", 19]]
  Spree::TaxCategory Load (0.4ms)  SELECT  "spree_tax_categories".* FROM "spree_tax_categories" WHERE "spree_tax_categories"."deleted_at" IS NULL AND "spree_tax_categories"."id" = $1 LIMIT 1  [["id", 1]]
   (0.5ms)  SELECT COUNT(*) FROM "spree_inventory_units" WHERE "spree_inventory_units"."line_item_id" = $1  [["line_item_id", 2]]
   (0.2ms)  COMMIT
```

These two dumps actually hint at more potential for optimization: The default zone really should be cached, but that can cause thread-safety issues. It'd need something like the `request_store` gem to get around them, and that's another PR. 

These two dumps are for an order with two line items. For every one of them, the Product, the Variant, a Zone and a TaxCategory is loaded. I could shave some of that off by passing IDs for Zone and TaxCategory and only loading them if the order's zone is not the default zone (in which case no conversion is necessary). Could we move that to another PR though? This one is getting bigger and bigger...

mamhoff commented:
Btw, is there any way to increase the character-per-line limit to 100? With all of Spree's namespacing, 80 is often impossible to attain...

BenMorganIO commented:
I'm :+1: for the transaction. I'm a lover of them and they keep everything "in sync".

JDutil commented:
@mamhoff earlier today I increased it to 120 to help reduce that spam:
https://github.com/spree/spree/commit/a7a3e88c1d7c7e8a5248cfa44a20d3075f1985ee

JDutil commented:
@mamhoff disregard removing the @ tag I can amend that now during squash, but FYI in the future it's against our contributing guidelines which I'll update if it's not already there.

JDutil commented:
:shipit: :beers:

tvdeyen commented:
Great work guys. Thanks for all the feedback. And especially @mamhoff for his continuously work on this 🍻👍
